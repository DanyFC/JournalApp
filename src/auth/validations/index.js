export { default as loginValidation } from './loginValidations'
export { default as registerValidation } from './registerValidation'