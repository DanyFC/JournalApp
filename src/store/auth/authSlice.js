import { createSlice } from '@reduxjs/toolkit'

export const authSlice = createSlice({
  name: 'auth',
  initialState: {
    displayName: null,
    email: null,
    errorMessage: null,
    photoURL: null,
    status: 'checking', // INFO: 'checking', 'not-authenticated', 'authenticated'
    uid: null
  },
  reducers: {
    checkCredential: (state) => {
      state.status = 'checking'
    },
    login: (state, { payload }) => {
      state.displayName = payload.displayName
      state.email = payload.email
      state.errorMessage = null
      state.photoURL = payload.photoURL
      state.status = 'authenticated'
      state.uid = payload.uid
    },
    logout: (state, { payload }) => {
      state.displayName = null
      state.email = null
      state.errorMessage = payload ? payload : null
      state.photoURL = null
      state.status = 'not-authenticated'
      state.uid = null
    }
  },
})

export const { checkCredential, login, logout } = authSlice.actions