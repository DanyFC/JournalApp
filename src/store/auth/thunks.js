import { checkCredential, login, logout } from './'
import { loginWithEmailPassword, logoutFirebase, registerUserWithEmailPassword, signInWithGoogle } from '../../firebase/providers'
import { clearState } from '../journal'

export const startLoginWithEmailPassword = (user) => {
  return async (dispatch) => {
    dispatch(checkCredential())
    const result = await loginWithEmailPassword(user)
    if (!result.ok) {
      dispatch(logout(result.errorMessage))
      return
    }
    dispatch(login(result))
  }
}

export const startGoogleSignIn = () => {
  return async (dispatch) => {
    dispatch(checkCredential())
    const result = await signInWithGoogle()
    if (!result.ok) {
      dispatch(logout(result.errorMessage))
      return
    }
    dispatch(login(result))
  }
}

export const startCreatingUserWithEmailPassword = (newUser) => {
  return async (dispatch) => {
    dispatch(checkCredential())
    const result = await registerUserWithEmailPassword(newUser)
    if (!result.ok) {
      dispatch(logout(result.errorMessage))
      return
    }
    dispatch(login(result))
  }
}

export const startLogout = () => {
  return async (dispatch) => {
    await logoutFirebase()
    dispatch(clearState())
    dispatch(logout())
  }
}