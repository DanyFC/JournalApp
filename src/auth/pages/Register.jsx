import { Alert, Button, Grid, Link, TextField, Typography } from '@mui/material'
import { Link as RouterLink } from 'react-router-dom'
import { registerValidation } from '../validations'
import { startCreatingUserWithEmailPassword } from '../../store/auth'
import { useDispatch, useSelector } from 'react-redux'
import { useForm } from '../../hooks'
import { useMemo } from 'react'
import Auth from '../layout/Auth'

const formData = {
  displayName: '',
  email: '',
  password: '',
}

const Register = () => {
  const dispatch = useDispatch()

  const { errorMessage, status } = useSelector(state => state.auth)
  const isAuthenticating = useMemo(() => status === 'checking', [status])

  const { displayName, email, password, formState, errors, onInputChange, onHandleBlur, onSubmit } = useForm(formData, registerValidation, createAccount)

  function createAccount() {
    dispatch(startCreatingUserWithEmailPassword(formState))
  }

  return (
    <Auth title='Register'>
      <form onSubmit={onSubmit} aria-label='register-form'>
        <Grid container>

          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              fullWidth
              error={!!errors.displayName}
              helperText={errors.displayName ? errors.displayName : null}
              label='Name'
              onChange={onInputChange}
              onBlur={onHandleBlur}
              placeholder='your name'
              name='displayName'
              type='text'
              value={displayName}
            />
          </Grid>

          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              fullWidth
              error={!!errors.email}
              helperText={errors.email ? errors.email : null}
              label='Email'
              onChange={onInputChange}
              onBlur={onHandleBlur}
              placeholder='example@domain.com'
              name='email'
              type='email'
              value={email}
            />
          </Grid>

          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              fullWidth
              error={!!errors.password}
              helperText={errors.password ? errors.password : null}
              inputProps={{ 'aria-label': 'Password' }}
              label='Password'
              onChange={onInputChange}
              onBlur={onHandleBlur}
              placeholder='●●●●●●●●'
              name='password'
              type='password'
              value={password}
            />
          </Grid>

          <Grid container spacing={2} sx={{ mb: 2, mt: 1 }}>
            {errorMessage && (
              <Grid item xs={12}>
                <Alert severity='error'>{errorMessage}</Alert>
              </Grid>
            )}
            <Grid item xs={12}>
              <Button variant='contained' type='submit' disabled={isAuthenticating} fullWidth>
                Create account
              </Button>
            </Grid>
          </Grid>

        </Grid>

        <Grid container direction='row' justifyContent='end'>
          <Typography sx={{ mr: 1 }} >Do you already have an account?</Typography>
          <Link component={RouterLink} color='inherit' to='/auth/login'>
            Log in!
          </Link>
        </Grid>

      </form>
    </Auth>


  )
}
export default Register