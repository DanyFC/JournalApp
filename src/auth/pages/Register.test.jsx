/**
 * @jest-environment jsdom
*/
import { authSlice } from '../../store/auth'
import { configureStore } from '@reduxjs/toolkit'
import { fireEvent, render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { notAuthenticatedState } from '../../fixtures/auth'
import { Provider } from 'react-redux'
import Register from './Register'

const store = configureStore({
  reducer: {
    auth: authSlice.reducer
  },
  preloadedState: {
    auth: notAuthenticatedState
  }
})

/* eslint-disable no-undef */
const mockStartCreatingUserWithEmailPassword = jest.fn()

jest.mock('./../../store/auth/thunks', () => ({
  startCreatingUserWithEmailPassword: (formData) => () => mockStartCreatingUserWithEmailPassword(formData)
}))

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => (fn) => fn()
}))

describe('Tests in <Register />', () => {

  beforeEach(() => jest.clearAllMocks())

  test('Can be match snapshot.', () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Register />
        </MemoryRouter>
      </Provider>
    )

    expect(container).toMatchSnapshot()
  })

  test('Should call startCreatingUserWithEmailPassword to submit.', () => {
    const formData = {
      name: 'userTest',
      email: 'test@test.com',
      password: '123456',
    }
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Register />
        </MemoryRouter>
      </Provider>
    )
    const nameInput = screen.getByRole('textbox', { name: 'Name' })
    const emailInput = screen.getByRole('textbox', { name: 'Email' })
    const passwordInput = screen.getByLabelText('Password')
    const form = screen.getByLabelText('register-form')

    fireEvent.change(nameInput, { target: { name: 'displayName', value: formData.name } })
    fireEvent.change(emailInput, { target: { name: 'email', value: formData.email } })
    fireEvent.change(passwordInput, { target: { name: 'password', value: formData.password } })
    fireEvent.submit(form)


    expect(mockStartCreatingUserWithEmailPassword).toHaveBeenCalled()
  })
})
/* eslint-enable no-undef */
