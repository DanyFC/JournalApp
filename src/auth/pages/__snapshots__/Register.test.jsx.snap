// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <Register /> Can be match snapshot. 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-ro1yt6-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 box-shadow animate__animated animate__fadeInDown css-1ar3nvi-MuiGrid-root"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-1jcm1rz-MuiTypography-root"
      >
        Register
      </h5>
      <form
        aria-label="register-form"
      >
        <div
          class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1hzi63y-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r0:"
                id=":r0:-label"
              >
                Name
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r0:"
                  name="displayName"
                  placeholder="your name"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Name
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1hzi63y-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r1:"
                id=":r1:-label"
              >
                Email
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r1:"
                  name="email"
                  placeholder="example@domain.com"
                  type="email"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Email
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1hzi63y-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r2:"
                id=":r2:-label"
              >
                Password
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  aria-label="Password"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r2:"
                  name="password"
                  placeholder="●●●●●●●●"
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Password
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-h80lhz-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="submit"
              >
                Create account
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container css-1svu6jh-MuiGrid-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1fbouff-MuiTypography-root"
          >
            Do you already have an account?
          </p>
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-xdxgn1-MuiTypography-root-MuiLink-root"
            href="/auth/login"
          >
            Log in!
          </a>
        </div>
      </form>
    </div>
  </div>
</div>
`;
